packages/backend-lalr/src/Happy/Backend/LALR/ProduceCode.lhs

```diff
@@ -581,28 +587,133 @@ machinery to discard states in the parser...
+
+%  {-# NOINLINE happyExpListPerState #-}
+%  happyExpListPerState st = token_strs_expected
+%        where token_strs = " . str (show $ elems token_names')
+%              bit_start = st Prelude.* " . str (show nr_tokens)
+%              bit_end = (st Prelude.+ 1) Prelude.* " . str (show nr_tokens)
+%              read_bit = readArrayBit happyExpList
+%              bits = Prelude.map read_bit [bit_start..bit_end Prelude.- 1]
+%              bits_indexed = Prelude.zip bits [0..nr_tokens - 1]
+%               token_strs_expected = Prelude.concatMap f bits_indexed
+%               f (Prelude.False, _) = []
+%               f (Prelude.True, nr) = [token_strs Prelude.!! nr]
+
 >    produceExpListPerState
->       = produceExpListArray
->       . str "{-# NOINLINE happyExpListPerState #-}\n"
->       . str "happyExpListPerState st =\n"
->       . str "    token_strs_expected\n"
->       . str "  where token_strs = " . str (show $ elems token_names') . str "\n"
->       . str "        bit_start = st Prelude.* " . str (show nr_tokens) . str "\n"
->       . str "        bit_end = (st Prelude.+ 1) Prelude.* " . str (show nr_tokens) . str "\n"
->       . str "        read_bit = readArrayBit happyExpList\n"
->       . str "        bits = Prelude.map read_bit [bit_start..bit_end Prelude.- 1]\n"
->       . str "        bits_indexed = Prelude.zip bits [0.."
->                                        . str (show (nr_tokens - 1)) . str "]\n"
->       . str "        token_strs_expected = Prelude.concatMap f bits_indexed\n"
->       . str "        f (Prelude.False, _) = []\n"
->       . str "        f (Prelude.True, nr) = [token_strs Prelude.!! nr]\n"
->       . str "\n"
+>       =
+>       (renderDocDecs $
+>            (runIdentity $ evalStateT (produceExpListArray @DocExp) Map.empty)
+>         ++ [[noInlinePragmaD happy_exp_list_per_state_name, happy_exp_list_per_state_dec]])
+>       . nl
 >       where (first_token, last_token) = bounds token_names'
 >             nr_tokens = last_token - first_token + 1
 >
+>          --happyExpListPerState st = token_strs_expected
+>             happy_exp_list_per_state_name = "happyExpListPerState"
+>             happy_exp_list_per_state_dec =
+>               funD happy_exp_list_per_state_name [
+>                 clause [st_pat] (varE token_strs_expected_name) [
+>                   token_strs_dec,
+>                   bit_start_dec,
+>                   bit_end_dec,
+>                   read_bit_dec,
+>                   bits_dec,
+>                   bits_indexed_dec,
+>                   f_dec,
+>                   token_strs_expected_dec
+>                 ]
+>               ]
+>
+>             st_name = "st"
+>             st_var = varE st_name
+>             st_pat = varP st_name
+>
+>           --token_strs = elems token_names'
+>             token_strs_name = "token_strs"
+>             token_strs_dec = funD token_strs_name [clause [] token_strs_exp []]
```

Comment:
These single-clause functions seem to be very widespread, so I think it would be easier to read if you introduced a shortcut, e.g. `simpleFunD name body = funD name [clause [] body []]`.